llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -2352,6 +2359,78 @@ Instruction *ModuleAddressSanitizer::CreateAsanModuleDtor(Module &M) {
   return ReturnInst::Create(*C, AsanDtorBB);
 }
 
+void ModuleAddressSanitizer::instrumentSyclDeviceGlobals(IRBuilder<> &IRB,
+                                                         Module &M) {
+  auto &DL = M.getDataLayout();
+  SmallVector<GlobalVariable *, 8> GlobalsToRemove;
+  SmallVector<GlobalVariable *, 8> NewDeviceGlobals;
+  SmallVector<Constant *, 8> DeviceGlobalMetadata;
+
+  // Device global meta data is described by a structure
+  //  size_t device_global_size
+  //  size_t device_global_size_with_red_zone
+  //  size_t beg
+  StructType *StructTy = StructType::get(IntptrTy, IntptrTy, IntptrTy);
+
+  for (auto &G : M.globals()) {
+    if (isDeviceGlobalVariable(G) && hasDeviceImageScopeProperty(G)) {
```

Comment:
will device globals in https://github.com/intel/llvm/blob/626c121cf2594df368636286585628d1189b87aa/libdevice/sanitizer_utils.cpp#L28-L35 be instrumented here as well?

Reply:
Device global in libdevice shouldn't be affected, the Asan Instrument Pass will only go user's device image in device compilation phase. Does this align with your expectation? @zhaomaosu

Reply:
Yeah, device globals in libdevice will not be affected since they don't have any properties.

Reply:
add a comment that hasDeviceImageScopeProperty is used for excluding device globals in libdevice? It isn't very intuitive why hasDeviceImageScopeProperty is called here.

Reply:
> add a comment that hasDeviceImageScopeProperty is used for excluding device globals in libdevice?

hasDeviceImageScopeProperty is not used for excuding device globals in libdevice.
ASanPass only runs when user kernel code is compiling, and after it is compiled, it will link with libdevice.
So ASanPass doesn't modify libdevice's code.

Reply:
ok. my comment is just based on the previous comment.
Then why is hasDeviceImageScopeProperty called here?

Reply:
> ok. my comment is just based on the previous comment. Then why is hasDeviceImageScopeProperty called here?

The device globals without "device_image_scope" property is implemented by the Device USM by the SYCL runtime, so it's automatically supported by us.
The device globals with "device_image_scope" property is more like the true global variable in one module.

Reply:
a comment explaining what device globals are excluded by the function would more intuitive. But it is fine if you think it is unnecessary.

Reply:
Thanks, added a comment for this. BTW, I have moved the instrument code to sycl-post-link tool.

Reply:
Why did you move device global instrumentation in sycl-post-link?

Reply:
If there are more than one user source files, since we will add extra global variables to each source file, multi definition errors will be reported by llvm-link tool.

Reply:
can we still instrument in instrumentation pass and adds AsanDeviceGlobalMetadata with appending linkage. Than in sycl-post-link we only need to create __AsanDeviceGlobalCount

Reply:
please also check if clGetDeviceGlobalVariablePointerINTEL can get ride of __AsanDeviceGlobalCount

Reply:
Discussed with Wenju offline, I'll create another PR to implement new UR API to get device global pointer. Before that, we will keep current solution. When the new API is implemented, I'll move this instrument part back to AddressSanitizer pass.